src/syntaxtree.c

```diff
@@ -0,0 +1,2113 @@
+
+    return result;
+}
+
+Obj SyntaxTreeProd(Expr expr)
+{
+    Obj result;
+
+    result = NewSyntaxTreeNode("Prod", 3);
+
+    AssPRec(result, RNamName("left"), SyntaxTreeExpr( ADDR_EXPR(expr)[0]));
+    AssPRec(result, RNamName("right"), SyntaxTreeExpr( ADDR_EXPR(expr)[1]));
+
+    return result;
+}
+
+Obj SyntaxTreeInv(Expr expr)
+{
+    Obj result;
+
+    result = NewSyntaxTreeNode("Inv", 3);
+
+    AssPRec(result, RNamName("op"), SyntaxTreeExpr( ADDR_EXPR(expr)[0]));
+
+    return result;
+}
+
+Obj SyntaxTreeQuo(Expr expr)
+{
+    Obj result;
+
+    result = NewSyntaxTreeNode("Quot", 3);
+
+    AssPRec(result, RNamName("left"), SyntaxTreeExpr( ADDR_EXPR(expr)[0]));
+    AssPRec(result, RNamName("right"), SyntaxTreeExpr( ADDR_EXPR(expr)[1]));
+
+    return result;
+}
+
+Obj SyntaxTreeMod(Expr expr)
+{
+    Obj result;
+
+    result = NewSyntaxTreeNode("Mod", 3);
+
+    AssPRec(result, RNamName("left"), SyntaxTreeExpr( ADDR_EXPR(expr)[0]));
+    AssPRec(result, RNamName("right"), SyntaxTreeExpr( ADDR_EXPR(expr)[1]));
+
+    return result;
+}
+
+Obj SyntaxTreePow(Expr expr)
+{
+    Obj result;
+
+    result = NewSyntaxTreeNode("Pow", 3);
+
+    AssPRec(result, RNamName("left"), SyntaxTreeExpr( ADDR_EXPR(expr)[0]));
+    AssPRec(result, RNamName("right"), SyntaxTreeExpr( ADDR_EXPR(expr)[1]));
+
+    return result;
+}
+
+Obj SyntaxTreeIntExpr(Expr expr)
+{
+    Obj result;
+    Obj value;
+    Int siz;
+    UInt typ;
+    Int i;
+
+    result = NewSyntaxTreeNode("IntExpr", 1);
+
+    if(IS_INTEXPR(expr)) {
+        value = OBJ_INTEXPR(expr);
+    } else {
+        siz = SIZE_EXPR(expr) - sizeof(UInt);
+        typ = *(UInt *)ADDR_EXPR(expr);
+        value = C_MAKE_INTEGER_BAG(siz, typ);
+        for ( i = 0; i < siz/INTEGER_UNIT_SIZE; i++ ) {
+#if INTEGER_UNIT_SIZE == 2
+            C_SET_LIMB2(value, i, ((UInt2 *)((UInt *)ADDR_EXPR(expr) + 1))[i]);
+#elif INTEGER_UNIT_SIZE == 4
+            C_SET_LIMB4(value, i, ((UInt4 *)((UInt *)ADDR_EXPR(expr) + 1))[i]);
+#elif INTEGER_UNIT_SIZE == 8
+            C_SET_LIMB8(value, i, ((UInt8 *)((UInt *)ADDR_EXPR(expr) + 1))[i]);
+#else
+            #error unsupported INTEGER_UNIT_SIZE
+#endif
+        }
+        if (siz <= 8) {
+            value = C_NORMALIZE_64BIT(value);
+        }
+    }
+
+    AssPRec(result, RNamName("value"), value);
+
+    return result;
+}
+
+Obj SyntaxTreeTrueExpr(Expr expr)
+{
+    Obj result;
+
+    result = NewSyntaxTreeNode("TrueExpr", 1);
+
+    return result;
+}
+
+Obj SyntaxTreeFalseExpr(Expr expr)
+{
+    Obj result;
+
+    result = NewSyntaxTreeNode("FalseExpr", 1);
+
+    return result;
+}
+
+Obj SyntaxTreeCharExpr(Expr expr)
+{
+    Obj result;
+    Obj value;
+
+    /* TODO: How do I make a character literal? */
+    result = NewSyntaxTreeNode("CharExpr",1);
+    value = INTOBJ_INT(ADDR_EXPR(expr)[0]);
+    AssPRec(result, RNamName("result"), value);
+
+    return result;
+}
+
+Obj SyntaxTreePermExpr (Expr expr)
```

Comment:
well, since you can have expressions in the permutation entries, I want to be able to reflect that, i.e. what do you do with

```
for i in [1..5] do
  p := p * (i, i+5);
od;
```